lightningd/plugin_control.c

```diff
@@ -0,0 +1,135 @@
+		plugin_found = false;
+		list_for_each(&cmd->ld->plugins->plugins, p, list) {
+			if (paths_match(p->cmd, control_param)) {
+				plugin_found = true;
+				plugin_hook_unregister_all(p);
+				plugin_kill(p, "%s stopped by lightningd via RPC",
+						control_param);
+				break;
+			}
+		}
+		if (!plugin_found)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					   "Could not find plugin %s", control_param);
+	}
+	else if (streq(control_command, "start")) {
+		if (access(control_param, X_OK) == 0)
+			plugin_register(cmd->ld->plugins, control_param);
+		else
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						   "Could not access %s with execution permission",
+						   control_param);
+	}
+	else if (streq(control_command, "adddir")) {
+		if (access(control_param, F_OK) == 0)
+			add_plugin_dir(cmd->ld->plugins, control_param, true);
+		else
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						   "Could not open %s", control_param);
+	}
+	else if (streq(control_command, "reload")) {
+		plugins_add_default_dir(cmd->ld->plugins,
+				path_join(tmpctx, cmd->ld->config_dir, "plugins"));
+	}
+	else if (streq(control_command, "unregister_hook")) {
+		plugin_found = false;
+		list_for_each(&cmd->ld->plugins->plugins, p, list) {
+			if (paths_match(p->cmd, control_param)) {
+				plugin_found = true;
+				if (plugin_hook_unregister(p, control_second_param))
+					break;
+				else
+					return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+							   "Could not unregister %s for plugin %s",
+							   control_param, control_second_param);
+			}
+		}
+		if (!plugin_found)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					   "Could not find plugin %s", control_param);
+	}
+	else if (streq(control_command, "list")) {
+		/* Don't do anything as we return the plugin list anyway */
+	}
+	else {
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					"Invalid control : %s should be one of"
+					" stop, start, adddir, reload",
+					control_command);
+	}
+
+	/* The config function is called once we got the manifest,
+	 * in 'plugin_manifest_cb'.*/
+	plugins_start(cmd->ld->plugins, cmd->ld->dev_debug_subprocess);
+
+	response = json_stream_success(cmd);
+	json_array_start(response, "active_plugins");
+	list_for_each(&cmd->ld->plugins->plugins, p, list) {
+		if (p->configured)
+			json_add_string(response, NULL, p->cmd);
+	}
+	json_array_end(response);
+
+	return command_success(cmd, response);
+}
+
+static const struct json_command plugin_control_command = {
+	"plugin",
+	"plugin",
+	json_plugin_control,
+	"Control plugins (start, stop, adddir, reload, list)",
+	.verbose = "Usage :\n"
+	"cli plugin start /path/to/a/plugin\n"
+	"	adds a new plugin to c-lightning\n"
+	"cli plugin stop plugin_name\n"
+	"	stops an already registered plugin\n"
+	"cli plugin adddir /path/to/a/plugin_dir/\n"
+	"	adds a new plugin directory\n"
+	"cli plugin reload\n"
+	"	loads not-already-loaded plugins from the default plugins dir\n"
```

Comment:
That is not what I expected to happen when I say `reload`; I expected it to take a plugin name to restart, or maybe `all` or nothing for everything?

Reply:
A plugin could be reloaded by `stop`ing and `start`ing it again, however I think it's a good design to involve the `plugins` directory and to make it possible to load all new plugins from this directory. Maybe the command name is not well chosen (as @ZmnSCPxj already pointed out)

Reply:
"rescandir"?